src/findlib.ml

```diff
@@ -298,273 +255,241 @@ let parse_package t ~name ~parent_dir ~vars ~required_by =
+
+  let not_available (na : Package_not_available.t) ~required_by =
+    raise(Findlib (Package_not_available { na with required_by }))
+
+  let deps_exn (deps : present_or_not_available list) ~required_by =
+    List.map deps ~f:(function
+      | Ok x -> x
+      | Error na -> not_available na ~required_by)
+
+  let requires_internal t =
+    match t.requires with
+    | Resolved ts -> ts
+    | Names names ->
+      let ts = List.map names ~f:(find_internal t.db) in
+      t.requires <- Resolved ts;
+      ts
+
+  let ppx_runtime_deps_internal t =
+    match t.ppx_runtime_deps with
+    | Resolved ts -> ts
+    | Names names ->
+      let ts = List.map names ~f:(find_internal t.db) in
+      t.ppx_runtime_deps <- Resolved ts;
+      ts
+
+  let requires t ~required_by =
+    deps_exn (requires_internal t) ~required_by
+  let ppx_runtime_deps t ~required_by =
+    deps_exn (ppx_runtime_deps_internal t) ~required_by
+end
+
+module Closure =
+  Top_closure.Make
+    (String)
+    (struct
+      type graph = unit
+      type t = Package.t * With_required_by.Entry.t list
+      let key (pkg, _) = pkg.name
+      let deps (pkg, required_by) () =
+        let pkgs = Package.requires_internal pkg in
+        let required_by =
+          With_required_by.Entry.Library pkg.name :: required_by
+        in
+        List.map pkgs ~f:(function
+          | Ok pkg   -> (pkg, required_by)
+          | Error na -> Package.not_available na ~required_by)
+    end)
 
-let check_deps_consistency ~required_by ~local_public_libs pkg requires =
-  List.iter requires ~f:(fun pkg' ->
+let check_deps_consistency ~required_by ~local_public_libs deps =
+  List.iter deps ~f:(fun pkg' ->
     match String_map.find pkg'.name local_public_libs with
     | None -> ()
     | Some path ->
       raise (Findlib (External_dep_conflicts_with_local_lib
                         { package             = pkg'.name
-                        ; required_by         = Library pkg.name
+                        ; required_by         = Library "TODO" (*pkg.name*)
                         ; required_locally_in = required_by
                         ; defined_locally_in  = path
                         })))
 
+let extend_error_stack e ~required_by =
+  match e with
+  | Package_not_available x ->
+    Package_not_available
+      { x with required_by = x.required_by @ required_by }
+  | External_dep_conflicts_with_local_lib x ->
+    External_dep_conflicts_with_local_lib
+      { x with required_locally_in = x.required_locally_in @ required_by }
+  | Dependency_cycle x ->
+    Dependency_cycle
+      { x with required_by = x.required_by @ required_by }
+
 let closure pkgs ~required_by ~local_public_libs =
-  remove_dups_preserve_order
-    (List.concat_map pkgs ~f:(fun pkg ->
-       check_deps_consistency ~required_by ~local_public_libs pkg pkg.requires;
-       pkg.requires)
-     @ pkgs)
+  match pkgs with
+  | [] -> []
+  | first :: others ->
+    let t = first.db in
+    let key =
```

Comment:
Wondering if this is such a good hash. If the libs are given in a different order than you must recalculate the closure again...

Reply:
Yh, that's the intent. If we hash over the sorted list of libraries, then the exact linking order will depend on unrelated things in the workspace, which doesn't seem right.